x-pack/plugins/security/server/deprecations/kibana_user_role.ts

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type {
+  SecurityGetRoleMappingResponse,
+  SecurityGetUserResponse,
+} from '@elastic/elasticsearch/api/types';
+
+import { i18n } from '@kbn/i18n';
+import type {
+  DeprecationsDetails,
+  DeprecationsServiceSetup,
+  ElasticsearchClient,
+  Logger,
+  PackageInfo,
+} from 'src/core/server';
+
+import type { SecurityLicense } from '../../common';
+import { getDetailedErrorMessage, getErrorStatusCode } from '../errors';
+
+export const KIBANA_USER_ROLE_NAME = 'kibana_user';
+export const KIBANA_ADMIN_ROLE_NAME = 'kibana_admin';
+
+interface Deps {
+  deprecationsService: DeprecationsServiceSetup;
+  license: SecurityLicense;
+  logger: Logger;
+  packageInfo: PackageInfo;
+}
+
+export const registerKibanaUserRoleDeprecation = ({
+  deprecationsService,
+  logger,
+  license,
+  packageInfo,
+}: Deps) => {
+  deprecationsService.registerDeprecations({
+    getDeprecations: async (context) => {
+      // Nothing to do if security is disabled
+      if (!license.isEnabled()) {
+        return [];
+      }
+
+      return [
+        ...(await getUsersDeprecations(context.esClient.asCurrentUser, logger, packageInfo)),
+        ...(await getRoleMappingsDeprecations(context.esClient.asCurrentUser, logger, packageInfo)),
+      ];
+    },
+  });
+};
+
+async function getUsersDeprecations(
+  client: ElasticsearchClient,
+  logger: Logger,
+  packageInfo: PackageInfo
+): Promise<DeprecationsDetails[]> {
+  let users: SecurityGetUserResponse;
+  try {
+    users = (await client.security.getUser()).body;
+  } catch (err) {
+    logger.error(`Failed to retrieve users: ${getDetailedErrorMessage(err)}`);
+    return deprecationError(packageInfo, err);
+  }
+
+  const usersWithKibanaUserRole = Object.values(users)
+    .filter((user) => user.roles.includes(KIBANA_USER_ROLE_NAME))
+    .map((user) => user.username);
+  if (usersWithKibanaUserRole.length === 0) {
+    return [];
+  }
+
+  return [
+    {
+      title: i18n.translate('xpack.security.deprecations.kibanaUser.deprecationTitle', {
+        defaultMessage: 'The "{userRoleName}" role is removed and "{adminRoleName}" role is added',
+        values: { userRoleName: KIBANA_USER_ROLE_NAME, adminRoleName: KIBANA_ADMIN_ROLE_NAME },
+      }),
+      message: i18n.translate('xpack.security.deprecations.kibanaUser.usersDeprecationMessage', {
+        defaultMessage:
```

Comment:
__note:__ I'd better suggest using custom roles with more precise Kibana privileges set, and only use `kibana_admin` as the last resort, but I couldn't figure out how to keep it short, concise and not confusing.